### PR TITLE
Prevents concurrency issues when refreshing access tokens

### PR DESCRIPTION
**Pull request by fperreaultnv**
GitHub Issue: #86 

## PR Type

What kind of change does this PR introduce?

- Bugfix

## Description

Using a `ConcurrentAuthenticationTokenProvider<AuthenticationData>` will allow the token to be refreshed only once, even if there are many parallel api calls.

## PR Checklist 
Please check if your PR fulfills the following requirements:

- [x] Interface members are XML documented
- [x] Documentation (XML or comments) has been added and/or existing documentation has been updated
- [x] [Architecture documents](./doc/Architecture.md) have been updated
- [ ] Tested on all relevant platforms

## Other information
Unable to do proper testing at the moment because the Test project does not build (see #87)

## Internal Issue (If applicable):
#86

### CODE REVIEW

**Comment by jeanplevesque** [comment]
> Note: By default the ConcurrentAuthenticationTokenProvider is not implemented in the template.
> Do we have any reasons to believe that we should not make it default?
> theoretically we should always have several calls from a VM trying to refresh their token.

Right, we should change the scaffolding code to leverage the `ConcurrentAuthenticationTokenProvider`.

**Comment by fperreaultnv** [comment]
> Right, we should change the scaffolding code to leverage the ConcurrentAuthenticationTokenProvider.

I think we first need to make the `ConcurrentAuthenticationTokenProvider` abstract in Mallard, this way, we could derive from it and implement the abstract methods `GetToken`, `NotifySessionExpired` and `RefreshToken` directly in application template. 

Doing this would make the implementation here a lot clearer.

**Comment by fperreaultnv** [comment]
See https://github.com/nventive/MallardMessageHandlers/pull/61
This would allow injection of IAuthenticationTokenHandler<TAuthenticationToken>

File: src/app/ApplicationTemplate.Shared/Business/Impl/AuthenticationService.cs [hunk 1]
```diff
@@ -1,28 +1,26 @@
...
-using System.Reactive.Subjects;
-using System.Reactive.Threading.Tasks;
 using System.Threading;
 using System.Threading.Tasks;
 using ApplicationTemplate.Client;
-using MallardMessageHandlers;
 
 namespace ApplicationTemplate.Business
 {
-	public partial class AuthenticationService : IAuthenticationService, IAuthenticationTokenProvider<AuthenticationData>
```
Comment (jeanplevesque) [inline src/app/ApplicationTemplate.Shared/Business/Impl/AuthenticationService.cs:14]: You can keep this recipe (`AuthenticationService` implementing `IAuthenticationTokenProvider<AuthenticationData>`) by using composition with `ConcurrentAuthenticationTokenProvider<AuthenticationData>`.

File: src/app/ApplicationTemplate.Shared/Business/Impl/AuthenticationTokenProvider.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+using System.Reactive.Subjects;
+using System.Threading;
+using System.Threading.Tasks;
+using ApplicationTemplate.Client;
+using MallardMessageHandlers;
+using Microsoft.Extensions.Logging;
+
+namespace ApplicationTemplate.Business
+{
+	public class AuthenticationTokenProvider : IAuthenticationTokenProvider<AuthenticationData>, INotifySessionExpired
```
Comment (jeanplevesque) [inline src/app/ApplicationTemplate.Shared/Business/Impl/AuthenticationTokenProvider.cs:14]: You don't need this new class if you keep the previous recipe.

File: src/app/ApplicationTemplate.Shared/Business/INotifySessionExpired.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+using System.Reactive;
+
+namespace ApplicationTemplate.Business
+{
+	/// <summary>
+	/// This interface allows to notify that a user session has expired through an observable sequence.
+	/// </summary>
+	public interface INotifySessionExpired
```
Comment (jeanplevesque) [inline src/app/ApplicationTemplate.Shared/Business/INotifySessionExpired.cs:9]: You don't need this new interface if you keep the previous recipe.

**Review by jeanplevesque — CHANGES_REQUESTED** [review]

**Review by jeanplevesque — APPROVED** [review]

**Review by takla21 — APPROVED** [review]

**Review by Iguins — APPROVED** [review]

**Comment by fperreaultnv** [comment]
@MatFillion @jeanplevesque I'd like to merge this, but I did not find the time to test it properly, and I'm not really set up to test this on physical devices. I also don't have a mac to test at all on iOS. What can we do?

**Comment by jeanplevesque** [comment]
@fperreaultnv There are many tests in Mallard for `ConcurrentAuthenticationTokenProvider` and it's being used in real apps. I'm confident we can merge this.

**Pull request merged**